### PR TITLE
Allow additional registries to be specified for the secret

### PR DESCRIPTION
**Pull request by Veraticus**
The returned AWS ECR credentials can potentially be valid for any registry; but, each registry must be specified in the Kubernetes secret for Kubernetes to know to use that secret for that registry.

This PR adds a new environment variable `DOCKER_REGISTRIES`. If set, each registry in that environment variable (comma-delimited) will be added to the docker config secret, in addition to the "default" registry returned via logging-in.

### CODE REVIEW

**Comment by nabsul** [comment]
@Veraticus: Thank you for contributing this feature. It might take me a week or two to fully review, test and release it, but overall it looks like an easy change that makes sense to add.

File: .github/workflows/dockerimage.yml [hunk 1]
```diff
@@ -24,7 +24,8 @@ jobs:
     - name: Push image
       run: |
         VERSION=$(date +%s)
-        IMAGE=docker.pkg.github.com/nabsul/k8s-ecr-login-renew/k8s-ecr-login-renew:$VERSION
+        REPOSITORY=$(echo "${GITHUB_REPOSITORY}" | tr '[:upper:]' '[:lower:]')
```
Comment (nabsul) [inline .github/workflows/dockerimage.yml:27]: Is this change intentional, or did it get added by mistake?

Comment (Veraticus) [reply .github/workflows/dockerimage.yml:27]: No, it was intentional; I've released this to my own private registry in order to use it while you're testing it, and this allows me to publish it to mine -- and, I assume, yours as well, or whoever else forks it in the future :)

File: main.go [hunk 2]
```diff
@@ -38,14 +40,20 @@ func main() {
 	checkErr(err)
 	fmt.Println("Success.")
 
+	servers := []string{server}
```
Comment (nabsul) [inline main.go:43]: sorry to be a bit nit-picky: I think this can be simplified to:

```
registries := strings.Split(os.Getenv(envVarRegistries), ",")
servers := append(registries, server)
```

File: src/k8s/kubernetes.go [hunk 3]
```diff
@@ -50,16 +61,17 @@ func getSecret(client *kubernetes.Clientset, name, namespace string) (*v1.Secret
...
+	auth := &Auth{
+		Username: username,
+		Password: password,
+		Email:    defaultEmail,
+		Auth:     base64.StdEncoding.EncodeToString([]byte(username + ":" + password)),
+	}
+
+	config := Config{Auths: make(map[string]*Auth, len(servers))}
+	for _, server := range servers {
+		config.Auths[server] = auth
```
Comment (nabsul) [inline src/k8s/kubernetes.go:74]: I was a little concerned this might fail if there were duplicate servers in the list, but this handles it just fine!

**Comment by nabsul** [comment]
Overall, the code changes look great. Clean, simple, and backwards compatible. I tested the code snippets and they work as expected. 

I do still want to test these changes locally, and review some AWS docs. Hopefully this weekend we can get this merged. 

Thanks again for this contribution!

File: main.go [hunk 4]
```diff
@@ -6,12 +6,14 @@ import (
...
 	"github.com/nabsul/k8s-ecr-login-renew/src/k8s"
 	"os"
+	"strings"
 	"time"
 )
 
 const (
 	envVarAwsSecret       = "DOCKER_SECRET_NAME"
 	envVarTargetNamespace = "TARGET_NAMESPACE"
+	envVarRegistries      = "DOCKER_REGISTRIES"
```
Comment (nabsul) [inline main.go:16]: Actually, now that I've thought about it a little more, I wonder if the list of registries is provided by the AWS API directly. That would be better than requiring users to manually enter a list....

**Comment by nabsul** [comment]
Hey @Veraticus, I've been looking at ECR registries, and I'm a little confused. Are you able to have multiple ECR registries? Or is it the same hostname, but in different regions, such as: 

- xxxxxx.dkr.ecr.us-west-2.amazonaws.com
- xxxxxx.dkr.ecr.us-east-2.amazonaws.com

I can't seem to find a way in AWS to create a new xxxxxxx - part of the registry, but I can create instances in different regions (but the xxxxx part stays the same).

**Comment by Veraticus** [comment]
Yes, you are; the xxxxxx is your account ID, so this is really to enable ECR access across multiple accounts.

**Comment by nabsul** [comment]
Oh wow. But can the same credentials work on registries in different accounts? That seem surprising to me.

I would have assumed that an IAM user on one AWS account could not have permission to access a registry in a different account....

**Comment by nabsul** [comment]
I'd like to make sure we add this feature in the best/easiest way possible. I'm looking at this documentation page: https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_GetAuthorizationToken.html

It's clear if we provide a list of registry IDs, the API should return multiple credential. But I'm confused because it also says that parameter is being deprecated.

So my next question is: If no registry IDs are provided, will it automatically return the full list of credentials?

**Comment by Veraticus** [comment]
Yes, as long as the user is allowed you can use the same credentials in registries across multiple accounts. That's our current production configuration. 😄 

Registry ID is deprecated because the credentials will work for any registry that the user has permissions to access. So you can provide any registry ID; as long as your user has permission to access it, the same credentials will work.

**Comment by nabsul** [comment]
But here's the plan if I'm understanding things correctly, it sounds like we should make the following two changes

- Your change: Adding the optional environment variable to add more registries
- My change: Reading all the authorizations from the AWS API, instead of just the first

I'm not familiar with multiple account setups, but I really want to be able to try this feature out before publishing an update. Would you be willing to provide me with instructions to set on an IAM user with access to two accounts? I've already created an org and a second account in AWS, but I'm not sure where to go after that....

**Comment by Veraticus** [comment]
> Reading all the authorizations from the AWS API, instead of just the first

I don't think you can enumerate all authorizations from the AWS API, though if you've found documentation otherwise you're welcome to try.

> Would you be willing to provide me with instructions to set on an IAM user with access to two accounts?

Sure! You need the following policy statement on your ECR:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::<IAM user's account ID>:root"
      },
      "Action": [
        "ecr:*"
      ]
    }
  ]
}
```

And on your IAM user:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "ecr:*"
            ],
            "Resource": [
                "arn:aws:ecr:us-west-2:<account ID of account with repository>:repository/<repository>"
            ]
        },
        {
            "Effect": "Allow",
            "Action": "ecr:GetAuthorizationToken",
            "Resource": "*"
        }
    ]
}
```

**Comment by nabsul** [comment]
Your instructions worked perfectly, thanks!

> I don't think you can enumerate all authorizations from the AWS API, though if you've found documentation otherwise you're welcome to try.

In the current code, I'm only taking the first entry of the array returned: https://github.com/nabsul/k8s-ecr-login-renew/blob/main/src/aws/aws.go#L17

I was hoping that the call to `GetAuthorizationToken()` would just automatically return an entry for each registry the user has access to, but it's looking like wishful thinking :-)

If you provide the list of registries it does exactly that, but since it's being deprecated this doesn't seem like a good solution.

It was an interesting trip digging through all this, but in the end I think your solution is the only way :-)

File: main.go [hunk 5]
```diff
@@ -38,14 +40,20 @@ func main() {
 	checkErr(err)
 	fmt.Println("Success.")
 
+	servers := []string{server}
+	registries := strings.Split(os.Getenv(envVarRegistries), ",")
```
Comment (nabsul) [inline main.go:44]: @Veraticus 

I have this fixed in the official 1.5 branch, but I thought I'd point it out so you don't end up making it elsewhere:

Splitting an empty string gives you an array of length 1, with one empty string in it. I think this is not the desired behavior here.

**Comment by PawelLipski** [comment]
@nabsul @Veraticus I've stumbled upon this brilliant project, that saved me a lot of work, thanks 🎉 

Actually, the missing feature I was thinking about is exactly what's done in this PR... would it be possible for you guys to make a pre-release build of this branch to be available on dockerhub? this would make my work even easier right now ^^

**Comment by PawelLipski** [comment]
Wow okay, I've seen `v1.5-rc5` 🥳

I've seen the custom-specific docker registry has been appended to the registry (completely useless in my case) that has been automatically inferred... is there going to be a way to completely skip the auto-inferred registry from the generated dockerconfigjson?

**Comment by nabsul** [comment]
Hi Pawel, I should be able to make an official release this weekend. There is `nabsul/k8s-ecr-login-renew:v1.5-rc5` if you want to try a prerelease. But as an RC it may be buggy.

> is there going to be a way to completely skip the auto-inferred registry

That's an interesting question. I think we should either keep the auto-inferred registry or always remove it. I don't want to add yet another switch somewhere for this.

Is there any good reason to remove the auto-inferred one?

**Comment by PawelLipski** [comment]
> Is there any good reason to remove the auto-inferred one?

Not sure about the other setups people might have, but in our case the auto-inferred is just a central account used for authentication, with ECR completely unused... so it just doesn't make sense to keep it around. I'd suggest that `DOCKER_REGISTRIES` should completely overwrite the inferred account -- even if someone wants to keep that inferred account for whatever reason, they can only add it explicitly to `DOCKER_REGISTRIES`.

**Comment by nabsul** [comment]
Hmmm... interesting points. To summarize my train of thoughts on this:

The question is: if DOCKER_REGISTRIES is set, should we:

a) Merge the list of registries with the default registry returned from AWS?
b) Ignore the default registry and only use the provided list (the user can of course re-add the default in the env var)

- Option b provides the most flexibility/control
- But option b also introduces a way for users to make a mistake when configuring the tool
- If you don't want the default registry of this account, why not create an IAM user id a different AWS account altogether?

I don't have strong opinions about this to be honest. I guess I'm slight in favor of option b (more flexibility, with a slight chance of confusing other users)

@Veraticus Do you have any thoughts on this suggestion?

**Comment by nabsul** [comment]
@PawelLipski nabsul/k8s-ecr-login-renew:v1.5-rc6 is up with your suggestion if you'd like to try it out.

**Comment by nabsul** [comment]
Either way, I'm pretty confident with the changes as they are. I'll just need to tidy up the docs and it should be ready for release some time this weekend.

**Comment by nabsul** [comment]
After updating the documentation, I noticed that the `TARGET_NAMESPACE` overrides the default namespace and does not merge with it. For me consistency is a great reason to stick with overriding in the case of `DOCKER_REGISTRIES`.

I'm going to go ahead and close this PR. v1.5 (https://github.com/nabsul/k8s-ecr-login-renew/pull/19) should be published shortly. Thank you both for your contributions.
